JP-3721: Simplify ModelContainer

emolter opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3721](https://jira.stsci.edu/browse/JP-3721)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #8738 

<!-- describe the changes comprising this PR here -->
With the addition of ModelLibrary, ModelContainer can be substantially simplified.  ModelLibrary certainly renders the `return_open` and `save_open` options obsolete, as well as the `get_sections` method, since if memory usage is a concern then ModelLibrary should be used instead.  Some additional discussion about how the ModelContainer used to perform way to many tasks can be found on [this innerspace page](https://innerspace.stsci.edu/pages/viewpage.action?spaceKey=SCSB&title=ModelContainer+uses+in+jwst).

During discussions related to [JP-3715](https://jira.stsci.edu/browse/JP-3715), it has become increasingly clear that the strictness of ModelLibrary and the additional borrow/shelve code required to access models is not necessary/desired for many use-cases.  For example, the `calwebb_spec3` pipeline currently makes extensive use of ModelContainer but does not have the same memory issues as `calwebb_image3` does.  Learning to use ModelLibrary may also be a nuisance for users manipulating relatively small datasets.

Therefore, ModelContainer should not be removed, but instead become a lightweight, easy-to-use class for loading a list of models and association metadata.  This PR proposed a container satisfying the following constraints:

- Loads association metadata dictionary
- Can be manipulated as if it were a list
- Loads all datamodels into memory (i.e., use ModelLibrary instead if we care about memory usage)
- Is the default data structure loaded by `datamodels.open()` for asn-type data but is no longer itself a datamodel
- Can be used as a context manager, i.e., `with datamodels.open(asn.json) as container` has the expected behavior

Linked PR in stdatamodels: https://github.com/spacetelescope/stdatamodels/pull/330 
Linked PR in stpipe: https://github.com/spacetelescope/stpipe/pull/190

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

emolter commented:
[Initial round of regression tests](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1754/#showFailuresLink) still showing several failures to track down

emolter commented:
The most recent changes [fixed most of those failures](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1757/).  There is still a lot to do, including docs updates and the like, but I think it's time to get some feedback on the overall idea behind the new ModelContainer, so I'm requesting a review from some folks now even though this is staying in draft mode.

perrygreenfield reviewed (commented):
I only had minor comments so far. In the end documentation on what the new ModelContainer class does and supports with reference to the changes relative to the older version should be provided.

emolter commented on jwst/datamodels/tests/data/association.json:
```diff
@@ -13,12 +13,12 @@
                 {
                     "exptype": "SCIENCE",
-                    "expname": "test.fits",
+                    "expname": "test.fits"
```
For my json linter, at least, yes.  I don't think these are being removed in Python

emolter commented:
> I only had minor comments so far. In the end documentation on what the new ModelContainer class does and supports with reference to the changes relative to the older version should be provided.

Thanks for the review Perry.  I agree we should add some documentation.  I'm still wondering what people think of the new-look container overall - is it complementary to ModelLibrary, does it make sense as a data structure, do any methods still need to be removed?

emolter commented:
After lots of discussion, it's been decided that this PR should attempt to avoid needing to make any stpipe changes.  This requires, among other things, putting back the `save` method.  Regression tests with that change and stpipe pinned to main [started here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1784/), to figure out what problems the current ModelContainer has when attempting to interface with stpipe/main.

melanieclarke reviewed (commented):
This seems like a pretty straightforward fix to me, now.  I want to try testing out a few more things in the interface before we move forward, but I have a couple quick comments now.

melanieclarke commented:
Also, one more question - is there a reason not to simplify the save function to just iterating over the models and calling model.save?

emolter commented on jwst/master_background/master_background_step.py:
```diff
@@ -154,19 +152,18 @@ def process(self, input):
-                    record_step_status(result, 'master_background', success=False)
-                    return result
+                    record_step_status(input_data, 'master_background', success=False)
+                    return input_data
```
I am not sure what difference it makes, but I am happy to remove the line that closes the model.

melanieclarke replied:
If you try to use a closed data model in subsequent steps, the pipeline will crash.  It seems unlikely this clause is ever encountered except in off-nominal situations, but we should probably still try to return a usable model.

Actually, looking again, I'm not sure why this clause needs modification at all - returning a copy of the input with the step status updated seems like the appropriate thing to do.

emolter replied:
you're right, I don't know why I changed that.  It is reverted now

melanieclarke commented on jwst/datamodels/container.py:
```diff
@@ -530,7 +464,15 @@ def get_crds_parameters(self):
+        to pass through step.run(), but it is never accessed.
         """
+        msg = ("stpipe uses the get_crds_parameters method from the 0th model in the "
+               "ModelContainer. This method is currently not used.")
+        raise NotImplementedError(msg)
         with self._open_first_science_exposure() as model:
             return model.get_crds_parameters()
 
```
The code below the NotImplementedError is unreachable.  If we want to keep it around in case it's useful in the future, maybe just warn and remove the NotImplementedError?  If we don't think the implementation is right at all, we should just remove it.

melanieclarke replied:
If we do keep `_open_first_science_exposure`, it needs a small fix to remove the self.meta reference.

melanieclarke replied:
Sorry, looking at it more - `_open_first_science_exposure` just looks fragile and unusable as is.  I say we remove it and just keep the NotImplementedError.

melanieclarke commented on docs/jwst/stpipe/devel_io_design.rst:
```diff
@@ -163,15 +163,50 @@ Many of the JWST calibration steps and pipelines expect an
+* The `ModelContainer` class is still a list-like object.
+* The `ModelContainer` class is still the default class returned by stdatamodels `open()`.
+* The `ModelContainer` class can still be used as a context manager, such that `with open(asn_file.json)`
+  still works.
```
Should probably also add here that ModelContainer can no longer hold a list of ModelContainers.

emolter replied:
added this, lmk if it looks ok

melanieclarke replied:
Looks good, thanks!

melanieclarke commented on jwst/cube_build/data_types.py:
```diff
@@ -89,7 +89,7 @@ def __init__(self, input, single, output_file, output_dir):
             self.input_models = input_models
             if not single:  # find the name of the output file from the association
-                self.output_name = input_models.meta.asn_table.products[0].name
+                self.output_name = input_models.asn_table["products"][0]["name"]
```
This line will crash with a KeyError if CubeBuildStep is called on a ModelContainer made from a list of models or filenames, instead of from an association.  It looks like there's a similar crash for the implementation on main as well, but it might be useful instead to add a function to ModelContainer to robustly get ASN metadata or throw a better error if not possible.

emolter replied:
I tried fixing this by making the default asn_table contain the correct keys, but all with None values.  I haven't tested this yet but will do so.

melanieclarke replied:
This gets further than the previous version, but still crashes.

I'm testing using regression test data from nirspec/ifu:
```
import glob
from jwst import datamodels
from jwst.cube_build import CubeBuildStep

fn = glob.glob('jw01249*_cal.fits')
container = datamodels.ModelContainer(fn)
CubeBuildStep.call(container)   #  or CubeBuildStep.call(fn) - both crash the same way
```

It now crashes when trying to make an output name:
```
--> 216     suffix = self.output_name_base[self.output_name_base.rfind('_') + 1:]
    217     if suffix in ['g140m', 'g235m', 'g395m', 'g140h', 'g235h', 'g395h', 'prism']:
    218         self.output_name_base = self.output_name_base[:self.output_name_base.rfind('_')]

AttributeError: 'NoneType' object has no attribute 'rfind'
```
If it's not easy to make a real default output name here, we can just catch the None when the output_filename is set and throw an error saying that ASN metadata is required.

emolter replied:
I tested this in the same way on the most recent commit and it runs successfully, feel free to try it yourself though

melanieclarke replied:
That's working now, thanks!

melanieclarke commented on jwst/master_background/master_background_step.py:
```diff
@@ -252,12 +253,11 @@ def split_container(container):
+    merge_tree(science.asn_table, container.asn_table)
     # Prune the background members from the table
-    for p in science.meta.asn_table.instance['products']:
+    for p in science.asn_table['products']:
```
Current default values for the asn_table might do bad things here too - maybe make members an empty list?

emolter replied:
will test this

melanieclarke reviewed (commented):
A few comments below, relating to the default structure for the asn_table.

emolter commented:
@melanieclarke Thanks for your initial testing of this.  I did some additional testing with the same setup and found the following:

> I'm testing using regression test data from nirspec/ifu:

It appears this test also fails on main, at the line `self.output_name = input_models.meta.asn_table.products[0].name` with an "AttributeError: No attribute 'asn_table'".  I was able to make the step at least run by getting the datatypes right in the EMPTY_ASN_TABLE, but this causes the output names to look strange: they start with an underscore, e.g., `_g395h-f290lp_s3d.fits`.  Of course this is worth fixing, but I'm not sure what's desired here because I can't get the step to run on a list of models on main.  Also, having never looked at CubeBuild in detail before, I have no idea how the file naming works, and it looks complicated with the whole DataTypes class and everything.

MasterBackgroundStep was indeed causing problems because a list of input models doesn't have dedicated "background" exposures, so I added a graceful exit (step skip with warning) in that case.  Filenames look fine but do not contain any association ID.

OutlierDetectionStep appears to work fine on both the list of files and the ModelContainer instantiated from them after one small change to the logic of that step's `guess_mode` function.  Filenames look fine but do not contain any association ID.

PixelReplaceStep appears to work fine on both the list of files and the ModelContainer instantiated from them, but the output filenames look like e.g. `step_PixelReplaceStep_7_pixelreplacestep.fits`.

RE changes to `save()`, stpipe expects to be able to pass a `save_model_func` into the save method, and since we can't change the way `stpipe` handles things, it's necessary to keep that here.  I still slightly simplified the method and added a unit test that shows the intended behavior.  But let me know if you think that behavior should still be changed.

I do plan to test the rest of these on main also, but I suspect at least some of these issues existed before.

I'm starting to wonder whether initializing a ModelContainer from a list should issue a warning or even be disallowed entirely - the whole point of ModelContainer is to hold asn metadata.

melanieclarke commented:
> I do plan to test the rest of these on main also, but I suspect at least some of these issues existed before.

Sorry if I was unclear - there are definitely already some issues on main with running multi-input steps on a list instead of an association, including the crash in cube_build.  I was just thinking that maybe we could start sorting them out here, or at least make sure that the design of ModelContainer doesn't get in the way of fixing them, especially if we're going to claim in the docs that "a `list` is a perfectly valid input and output to a pipeline or step".  

> **I'm starting to wonder whether initializing a ModelContainer from a list should issue a warning or even be disallowed entirely** - the whole point of ModelContainer is to hold asn metadata.

I would prefer to keep the ability to make a container from a list, if at all possible. Right now, feeding a list of filenames to cube_build makes a ModelContainer anyway, which makes sense so that the rest of the code can treat the input consistently.  Making an asn file is a lot of overhead, sometimes it's nice to just be able to run a step on a list of input.

For less weird output filenames, we could maybe set the asn name to the basename for the first member model.  But I don't think we need to have a complete answer now - these things don't currently work on main anyway, so if it's possible to support them, we can gather more input on what the output naming conventions should be. If it just doesn't make sense to support running these steps without an association, we should just throw an exception to make that clear to the user.

emolter commented:
> "a list is a perfectly valid input and output to a pipeline or step"

This comment was meant primarily for developers, as in, "when deciding how a step is supposed to work, remember that you have the option to pass a list in and out of steps, you don't always need to turn it into a container".  I can clarify this.

> For less weird output filenames, we could maybe set the asn name to the basename for the first member model

Seems like a good idea, will try.

> I don't think we need to have a complete answer now - these things don't currently work on main anyway

Good to know where the goalpost is.  If you're ok with it, I'm going to try the default asn name thing, and then just make it so all four of the steps I tested above run on both a list-initialized container and a list of filenames.  Then I'll open a separate issue or issues for additional issues I notice.

Regression tests are all passing except the two unrelated tweakreg catalog ones.

braingram commented on jwst/datamodels/container.py:
```diff
@@ -315,36 +280,27 @@ def from_asn(self, asn_data):
-        # Pull the whole association table into meta.asn_table
-        self.meta.asn_table = {}
+        # Pull the whole association table into the asn_table attribute
+        self.asn_table = {}
```
```suggestion
        self.asn_table = copy.deepcopy(asn_data)
```
Would this work to remove the `properties.merge_tree` usage?

emolter replied:
this does appear to work, will update

emolter replied:
This is removed now, lmk how it looks @braingram

melanieclarke commented:
> This comment was meant primarily for developers, as in, "when deciding how a step is supposed to work, remember that you have the option to pass a list in and out of steps, you don't always need to turn it into a container". I can clarify this. Although it seems like that might be a bad idea currently because of the way stpipe tries to save step output...

Yeah, maybe we should just remove that line for now.

> Good to know where the goalpost is. If you're ok with it, I'm going to try the default asn name thing, and then just make it so all four of the steps I tested above run on both a list-initialized container and a list of filenames. Then I'll open a separate issue or issues for additional issues I notice.

Sounds good, thanks!

melanieclarke commented on jwst/datamodels/container.py:
```diff
@@ -359,61 +315,47 @@ def from_asn(self, asn_data):
         for idx, model in enumerate(self):
-            if len(self) <= 1:
-                idx = None
+            print(model)
+            print(model.meta.filename)
+            print(save_model_func)
```
Otherwise, this version of the save function looks fine to me - testing it with path, dir_path, and save_model_func options seems to work as expected.

emolter commented:
The suggested addition of a default name for list-like input seems to fix the file name issues, or at least make the output names less silly-looking, for all but the pixel replace step.  I opened an issue for that one: https://github.com/spacetelescope/jwst/issues/8904. 

The following script now runs successfully for both a list of filenames and a container initialized with a list of filenames:
```
# data from test_bigdata/jwst-pipeline/dev/nirspec/ifu/
fn = glob.glob(stem + "jw01249*_cal.fits")
container = datamodels.ModelContainer(fn)
# container = fn

CubeBuildStep.call(container, save_results=True)
MasterBackgroundStep.call(container, save_results=True)
OutlierDetectionStep.call(container, save_results=True)
PixelReplaceStep.call(container, save_results=True, skip=False)
```

melanieclarke reviewed (approved):
This all looks good to me now - thanks for all the updates!

We should probably run regression tests one more time, but I'll approve from my perspective now.

emolter commented:
Tracking down unit test failures now, all of which seem to have been introduced by my most recent commit.

emolter commented:
[new regtests started here](https://github.com/spacetelescope/RegressionTests/actions/runs/11408380018)

tapastro reviewed (approved):
👍
